### ISSUE TITLE
Add job label/display name

### ISSUE THREAD

**Issue by Snaipe** [travis-ci/travis-ci#5898]
The build summary screen can become a hassle to read for projects with a complex build matrix.

It would be great if we could add names/labels to specific jobs in the matrix.

The visual result could be something like this:

![travis-job-label-request](https://cloud.githubusercontent.com/assets/977973/14487767/d3fbf2e0-0164-11e6-82f9-82d1aa95629c.png)

And the .travis.yml config, like so:

``` yaml
matrix:
  include:
  - name: Build with foo as bar
```

Having no display name would simply blank out the relevant section.

This, if it were to be implemented:
- Would allow developers to see at a glance which builds failed an succeeded -- currently, I have to count the index of the job or read its log to see that a job cross compiles for architecture X or Y.
- Would allow developers to sum up the contents of the environment field to the right, which can become easily cluttered when you have 3+ variables defined in your matrix.
- Would allow developers to make a distinction between otherwise visually identical builds (e.g. builds that only differ by a secure variable, like builds testing on two specialized remotes machines with credentials, or builds that have the same environment variables but use different apt packages).

A possible workaround could be putting a `_` environment variable -- with the name as a value -- at the start of the environment list (not tested), but this is hardly convenient and is visually inconsistent.

**Comment by jimdebeer** [comment]
yes please!

**Comment by BanzaiMan** [comment]
This attribute, if implemented, is meaningful only for those that you explicitly include in `matrix.include` (because others are created as a cartesian product of matrix dimensions). In such cases, though, you can explicitly define just as easily an environment variable which will show up in the current display.

```yaml
matrix:
  include:
    - env: NAME="Build with foo as far" VAR1=VAL1 VAR2=VAL2
```

**Comment by jimdebeer** [comment]
Its only a cosmetic feature --- since env vars don't really look nice 

see the difference
![image](https://cloud.githubusercontent.com/assets/5682524/17127981/26463246-5312-11e6-93c5-ac9da820fe0c.png)
look pretty!

![image](https://cloud.githubusercontent.com/assets/5682524/17127987/33328cca-5312-11e6-9cd8-2c105e375266.png)
looks ugly!

**Comment by jayvdb** [comment]
While this may help individual teams, I think it detracts from the travis community, as it means there is no longer a common logic for understanding a build matrix screen. If it is done, I hope it is togglable, so someone can switch back to raw envvars.

Arch is already a separate column shown with an icon. Maybe add a variant icon for 32 vs 64 is needed?

IMO debug vs release should be a builtin travis feature (are there any supported languages where tbis doesnt make sense?) and shown in another column using an icon. Not sure about deploy.

**Comment by Snaipe** [comment]
@jayvdb 

> While this may help individual teams, I think it detracts from the travis community, as it means there is no longer a common logic for understanding a build matrix screen. If it is done, I hope it is togglable, so someone can switch back to raw envvars.

It should of course be completely optional. This is only relevant for non-trivial, hand-crafted build matrices (typically on multi-arch & cross-platforms projects w/ assembly)

> Arch is already a separate column shown with an icon. Maybe add a variant icon for 32 vs 64 is needed?

Arch does not suffice by itself because of cross-compilation.

Here's my build matrix: https://travis-ci.org/Snaipe/Mimick/builds/141036082

Can you tell which builds are i386, x86_64, AArch64, ARM? Not being able to name jobs for what they simply are is a hassle. I am reluctant to add more architectures to the CI config precisely because of this.

> IMO debug vs release should be a builtin travis feature (are there any supported languages where tbis doesnt make sense?) and shown in another column using an icon. Not sure about deploy.

I don't see working in practice, as this is highly dependent on both the language and the build system used. How would you make that work for projects using autotools for instance, where the debug mode is a user-defined flag?

**Comment by jayvdb** [comment]
As you are cross compiling, that should be in the Compiler column of Travis build matrix.  Currently it says "gcc", but that isnt really true.  The same is roughly true of multi-arch platforms.

Wrt [your config](https://github.com/Snaipe/Mimick/blob/master/.travis.yml) specifically, wouldnt it be enough if Travis build env column said "CONFIGURATION=Debug, *i386-env", in which case it is a minor change to the UI to better show complex yaml.  Then you could defined `x86_64-linux-debug` and it would appear as "*x86_64-linux-debug" in the env column -- at least then it isnt an arbitrary name; it is a defined name elsewhere in the yaml.

As for debug, on most languages, it is quite easily set with an env var.
For C/C++, it is CFLAGS and CXXFLAGS respectively, and the usual approach is to use `-g` or `-O2`.  autoconf understands those.  With direct clang invocation, I think `llvm-config` needs to be used, but I am rusty on that. The build script could still fiddle with those defaults env vars for project specific preferences wrt debug/release, but it isnt hard to have sane defaults which tease out bugs. (and then it would be possibly to add support for fancy config, such as building only debug versions by default for branches (so it is quicker), and enable release builds on master.

**Comment by swarren** [comment]
We're hitting this issue in U-Boot too, and the proposed feature is exactly what I was searching for. It seems like a good generic fix for the issue to me.

Naming jobs as an implicit side-effect of the environment variables that are set seems like a hack to me, especially if I wanted to just run different scripts instead rather than setting environment variables to influence the operation of a different script.

For reference, the WAR I applied to U-Boot's Travis config:
https://patchwork.ozlabs.org/patch/686225/
[U-Boot,1/2] travis-ci: set env vars to name jobs

**Comment by ProfessorKaos64** [comment]
What I really am looking for, is the ability to have labels for each env outputted as a clean label. Instead of just "Build stats: passing" I want sub labels for each job/env that runs. This would be ideal for folks who run multiple passes. My use case here is [multi-distro builds](https://github.com/ProfessorKaos64/ges-code/blob/develop-master/.travis.yml) via Docker images.

**Comment by joshk** [comment]
I like the idea of supporting this, and would be interested in adding it when explicitly listing jobs and their configurations.

I'll be adding this to our Roadmap for Q1 of next year.

**Comment by JohannesBeranek** [comment]
@joshk did anything happen on this?

Another reason an env var doesn't always work: As soon as env vars are different, the cache can't be shared anymore (which IMHO is a different issue - would be good to ignore certain things for cache dependency or define the exact cache dependency)

**Comment by simonvanderveldt** [comment]
> Another reason an env var doesn't always work: As soon as env vars are different, the cache can't be shared anymore (which IMHO is a different issue - would be good to ignore certain things for cache dependency or define the exact cache dependency)

This has just tripped me up. Eventually I noticed the mention of it in the cache docs, but I totally didn't expect env vars to have an impact on which cache is shared.

This means I'll have to remove those env vars and I only used them to identify what each job in the matrix was doing, so now I'm back to a list that doesn't really tell me anything about what each job does :(

**Comment by gnzlbg** [comment]
It's 2018. Any updates on this? @BanzaiMan  ?

**Comment by shs96c** [comment]
Adding to this: without job labels, this build provides very little actionable information unless someone delves into the outputs themselves:

https://travis-ci.org/SeleniumHQ/selenium/builds/331260746

We don't want to use environment variables since the cache warming stage can be expensive and is slow, but without them it's impossible to tell which of these are the JS tests, and which ones are the Java tests (or which browser is being used)

**Comment by gr2m** [comment]
As a workaround, we do this

```yaml
script: npm run $COMMAND
jobs:
  include:
    - stage: test
      env: COMMAND=test:unit
    - env: COMMAND=test:integratio
```

instead of this

```yaml
jobs:
  include:
    - stage: test
      script: npm run test:unit
    - env: npm run  test:integratio
```

Example: https://travis-ci.org/octokit/fixtures/builds/336419815

**Comment by azu** [comment]
I have similar problem on Travis CI API.
https://developer.travis-ci.com/resource/job#Job
API does not includes `env` variable.
I want to get display name or `env` value from job id.

**Comment by wking** [comment]
The Travis GitHub App [Checks support][1] includes a table with columns for matrix variables.  For an example with a Rust-version column, see [here][2].  I don't know if changing `script` will create a new column or not.

[1]: https://blog.travis-ci.com/2018-05-07-announcing-support-for-github-checks-api-on-travis-ci-com
[2]: https://github.com/rust-lang-nursery/license-exprs/pull/24/checks

**Comment by plata** [comment]
I'm currently using build stages to mitigate this issue, i.e. I put single jobs into their own build stage to name them. Of course, this is not really a solution because it means that the jobs cannot be executed in parallel anymore. However, the alternative is that people do not know what's wrong with their pull requests, so I'm willing to accept that issue for now.

**Comment by jakul** [comment]
Travis now supports naming jobs within a build matrix - https://docs.travis-ci.com/user/customizing-the-build/#naming-jobs-within-matrices

**Comment by joshk** [comment]
@swarren we are working on this right now, and hope to have something ready in the next week or two.

In general, since Job naming is supported, can we close this issue?

**Comment by swarren** [comment]
Yes, I'm fine with closing this if you open another issue to cover displaying the job name within the job page itself. Note that I'm not the original issue requester here though.

**Comment by Snaipe** [comment]
I'm fine with this being closed. Thank you for addressing this. :)

**Comment by joshk** [comment]
@swarren we should have this deployed in the next week or two max

**Comment by joshk** [comment]
@Snaipe thanks for closing the issue :)

**Comment by drdavella** [comment]
This is awesome, thanks for the work! Is there any chance you'd be able to continue displaying the environment variables in addition to the name? It seems like in most cases there's plenty of room.

**Comment by giampaolo** [comment]
How do we use this? What's the directive name? `name`?
